python/GafferSceneUI/VisualiserToolUI.py

```diff
@@ -118,3 +125,83 @@
 
 	},
 )
+
+class _DataNameChooser( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug, **kw ) :
+
+		self.__menuButton = GafferUI.MenuButton(
+			text = plug.getValue(),
+			menu = GafferUI.Menu( Gaffer.WeakMethod( self.__menuDefinition ) )
+		)
+
+		GafferUI.PlugValueWidget.__init__( self, self.__menuButton, plug, **kw )
+
+	def __menuDefinition( self ) :
+
+		menuDefinition = IECore.MenuDefinition()
+
+		node = self.getPlug().node()
+		if not isinstance( node, GafferSceneUI.VisualiserTool ) :
+			return
+		if self.getPlug() != node["dataName"] :
+			return
+
+		scenePlug = node.view()["in"].getInput()
+		scriptNode = node.view().scriptNode()
+		with node.view().context() :
+			selection = GafferSceneUI.ScriptNodeAlgo.getSelectedPaths( scriptNode )
+
+			primVars = set()
+
+			for path in selection.paths() :
+				if not scenePlug.exists( path ) :
+					continue
+
+				primitive = scenePlug.object( path )
+				if not isinstance( primitive, IECoreScene.MeshPrimitive ) :
+					continue
+
+				for v in primitive.keys() :
+					if primitive[v].interpolation not in [
+						IECoreScene.PrimitiveVariable.Interpolation.FaceVarying,
+						IECoreScene.PrimitiveVariable.Interpolation.Uniform,
+						IECoreScene.PrimitiveVariable.Interpolation.Vertex,
+					] :
+						continue
+
+					if not isinstance(
+						primitive[v].data,
+						(
+							IECore.IntVectorData,
+							IECore.FloatVectorData,
+							IECore.V2fVectorData,
+							IECore.Color3fVectorData,
+							IECore.V3fVectorData,
+						)
+					) :
+						continue
+
+					primVars.add( v )
+
+		if len( primVars ) == 0 :
+			menuDefinition.append( "/None Available", { "active" : False } )
+
+		else :
+			for v in reversed( sorted( primVars ) ) :
+				menuDefinition.prepend(
+					"/" + v,
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__setDataName ), v ),
+						"checkBox" : self.getPlug().getValue() == v,
+					}
+				)
+
+		menuDefinition.prepend( "/PrimVarDivider", { "divider" : True, "label" : "Primitive Variables" } )
+
+		return menuDefinition
+
+	def __setDataName( self, value, *unused ) :
+
+		self.getPlug().setValue( value )
+		self.__menuButton.setText( value )
```

Comment:
This should really be done in an `_updateFromValues()` override, so that the button is updated no matter how the plug's value is changed.

Reply:
Fixed and squashed into https://github.com/GafferHQ/gaffer/pull/6181/commits/c78be9e302b870aec9724a4263d0765be2bd12cd. I guess using the `---` string in case someone tries to assign multiple different values to the plug is appropriate? I'm not sure it entirely makes sense, maybe we should error in that rare case?

Reply:
You'll only get multiple different values when the widget has been configured to edit multiple plugs at once, which won't happen for this widget. But I think an empty string will now be replaced with `---`, which probably doesn't make sense? Maybe `None` would be better?

Reply:
While I'm nitpicking, let's also move `_updateFromValues()` above the private methods.

Reply:
That sounds good. Changed to `None` and moved the method above the private methods in https://github.com/GafferHQ/gaffer/pull/6181/commits/dfe1c9ac7e885e9fe94a4922b787b852a9978e1b